File: server/bindings.py
"""
Types for the server
Created by sheepy0125 and Brushfire257
MIT License | 2023-02-20
"""

### Setup ###
# Imports
from json import dumps


### Helper functions ###


def unsigned_float(value: float | int | str) -> float:
    """A non-negative float value.

    :raises ServerException: If the value is not a valid float
    :raises ServerException: If the value is negative
    """

    if not isinstance(value, float):
        try:
            value = float(value)
        except Exception as e:
            raise ServerException(
                enum_variant=Error.MalformedResponseTypeError, inner=e
            )

    if value < 0.0:
        raise ServerException(
            enum_variant=Error.MalformedResponseTypeError,
            inner=TypeError(f"Negative unsigned float: {value}"),
        )

    return value


def unsigned_int(value: int | float | str) -> int:
    """A non-negative integer value.

    :raises ServerException: If the value is not a valid integer
    :raises ServerException: If the value is negative
    """

    if not isinstance(value, int):
        try:
            value = int(value)
        except Exception as e:
            raise ServerException(
                enum_variant=Error.MalformedResponseTypeError, inner=e
            )

    if value < 0:
        raise ServerException(
            enum_variant=Error.MalformedRequestTypeError,
            inner=TypeError(f"Negative unsigned integer: {value}"),
        )

    return value


### Classes ###


class ServerException(Exception):
    """A custom server exception"""

    def __init__(self, *args, enum_variant: int, inner: Exception):
        super().__init__(*args)

        self.enum_variant = enum_variant
        self.inner = inner


### Enums ###


class Enum:
    @classmethod
    def lookup_by_prefix(cls, prefix: str) -> int | None:
        ...

    @classmethod
    def lookup_by_variant(cls, variant: int) -> str | None:
        ...


class Error(Enum):
    """This enum is non serializable"""

    MalformedRequestFailedPrefixParsing = 0
    MalformedRequestFailedCommandParsing = 1
    MalformedRequestFailedSeparatorParsing = 2
    MalformedRequestFailedArgumentsParsing = 3
    MalformedRequestFailedMetadataParsing = 4
    MalformedRequestTypeError = 5
    MalformedRequestOtherError = 6
    MalformedResponseTypeError = 10
    MalformedResponseOtherError = 11
    FailedToStartAlreadyStarted = 21
    FailedToStartMagnetOdometerFailed = 22
    FailedToStartMotorControlFailed = 23
    FailedToStartCouldNotAcquireDistanceLock = 24
    FailedToStopNotStarted = 25
    FailedToStopStartThreadWouldNotRespond = 26
    FailedStatusCouldNotAcquireDistanceLock = 27
    FailedPingNegativeLatency = 27
    AnyOtherError = 99


class Direction(Enum):
    """This enum is non serializable"""

    Backward = -1
    Stopped = 0
    Forward = 1


class GPIOPin(Enum):
    """This enum is non serializable"""

    MotorControllerForwardPositive = 17  # GPIO_GEN0
    MotorControllerForwardNegative = 27  # GPIO_GEN2
    MotorControllerBackwardPositive = 23  # GPIO_GEN4
    MotorControllerBackwardNegative = 24  # GPIO_GEN5
    MagnetHallEffectSensor = 25  # GPIO_GEN6


class TransitMode(Enum):
    ClientToServerRequest = 0
    ServerToClientResponse = 1

    @classmethod
    def lookup_by_prefix(cls, prefix: str) -> int | None:
        """Lookup the transit mode's variant by its prefix

        :raises ValueError: If the prefix is not 1-lengthed
        """

        # Ensure the prefix is valid
        if len(prefix) != 1:
            raise ValueError("Prefix is not 1-lengthed")

        match prefix:
            case "?":
                return cls.ClientToServerRequest
            case "~":
                return cls.ServerToClientResponse
            case _:
                return None

    @classmethod
    def lookup_by_variant(cls, variant: int) -> str | None:
        match variant:
            case cls.ClientToServerRequest:
                return "?"
            case cls.ServerToClientResponse:
                return "~"
            case _:
                return None


class TransitType(Enum):
    Request = 0
    Response = 1

    @classmethod
    def lookup_by_prefix(cls, prefix: str) -> int | None:
        """Lookup the transit type's variant by its prefix

        :raises ValueError: If the prefix is not 1-lengthed
        """

        # Ensure the prefix is valid
        if len(prefix) != 1:
            raise ValueError("Prefix is not 1-lengthed")

        match prefix:
            case "?":
                return cls.Request
            case "~":
                return cls.Response
            case _:
                return None


class Command(Enum):
    Ping = 0
    Start = 1
    Stop = 2
    StaticStatus = 3
    Status = 4
    Unknown = Error = 99

    @classmethod
    def lookup_by_name(cls, name: str) -> int | None:
        """Lookup a command's variant by its name"""

        uppercase_name = name.upper()
        match uppercase_name:
            case "PING":
                return cls.Ping
            case "START":
                return cls.Start
            case "STOP":
                return cls.Stop
            case "STATICSTATUS":
                return cls.StaticStatus
            case "STATUS":
                return cls.Status
            case "ERROR" | "UNKNOWN":
                return cls.Error
            case _:
                return None

    @classmethod
    def lookup_by_variant(cls, variant: int) -> str | None:
        """Lookup a command's name by its variant"""

        match variant:
            case cls.Ping:
                return "PING"
            case cls.Start:
                return "START"
            case cls.Stop:
                return "STOP"
            case cls.Status:
                return "STATUS"
            case cls.StaticStatus:
                return "STATICSTATUS"
            case cls.Unknown:
                return "UNKNOWN"
            case _:
                return None


### Structs ###


class SerializableStruct:
    def __init__(self, *_args, **_kwargs):
        ...

    @property
    def __dict__(self) -> dict:
        return {}

    def __str__(self) -> str:
        return dumps(self.__dict__)


# Metadata


class MetaData:
    def __init__(self, time: float):
        self.time = unsigned_float(time)

    @property
    def __dict__(self) -> dict:
        return {"time": self.time}


# Error


class ErrorResponse(SerializableStruct):
    def __init__(self, error_variant: int, message: str):
        self.error_variant = error_variant
        self.message = message

    @property
    def __dict__(self) -> dict:
        return {"error_variant": self.error_variant, "message": self.message}


# Ping


class PingArguments(SerializableStruct):
    def __init__(self, time: float):
        self.time = time

    @property
    def __dict__(self) -> dict:
        return {"time": self.time}


class PingResponse(SerializableStruct):
    def __init__(self, sent_time: float):
        self.sent_time = sent_time

    @property
    def __dict__(self) -> dict:
        return {"sent_time": self.sent_time}


# Start


class StartArguments(SerializableStruct):
    def __init__(self, distance: float, reverse_brake: bool):
        self.distance = unsigned_float(distance)
        self.reverse_brake = bool(reverse_brake)

    @property
    def __dict__(self) -> dict:
        return {
            "distance": self.distance,
            "reverse_brake": self.reverse_brake,
        }


class StartResponse(SerializableStruct):
    ...


# Stop


class StopArguments(SerializableStruct):
    ...


class StopResponse(SerializableStruct):
    ...


# Static status


class StaticStatusArguments(SerializableStruct):
    ...


class StaticStatusResponse(SerializableStruct):
    def __init__(self, number_of_magnets: int, wheel_diameter: float):
        self.number_of_magnets = number_of_magnets
        self.wheel_diameter = wheel_diameter

    @property
    def __dict__(self) -> dict:
        return {
            "number_of_magnets": self.number_of_magnets,
            "wheel_diameter": self.wheel_diameter,
        }


# Dynamic (regular) status


class DistanceInformation(SerializableStruct):
    def __init__(
        self,
        distance: float,
        velocity: float,
        magnet_hit_counter: int,
    ):
        self.distance = unsigned_float(distance)
        self.velocity = unsigned_float(velocity)
        self.magnet_hit_counter = unsigned_int(magnet_hit_counter)

    @property
    def __dict__(self) -> dict:
        return {
            "distance": self.distance,
            "velocity": self.velocity,
            "magnet_hit_counter": self.magnet_hit_counter,
        }


class StatusArguments(SerializableStruct):
    ...


class StatusResponse(SerializableStruct):
    def __init__(
        self,
        running: bool,
        uptime: int,
        runtime: int,
        distance: DistanceInformation,
    ):
        self.running = bool(running)
        self.uptime = unsigned_int(uptime)
        self.runtime = unsigned_int(runtime)
        self.distance = distance.__dict__

    @property
    def __dict__(self) -> dict:
        return {
            "running": self.running,
            "uptime": self.uptime,
            "runtime": self.runtime,
            "distance": self.distance,
        }


### Lookup tables ###


COMMAND_SERIALIZABLE_LUT = {
    Command.Ping: {
        TransitType.Request: PingArguments,
        TransitType.Response: PingResponse,
    },
    Command.Start: {
        TransitType.Request: StartArguments,
        TransitType.Response: StartResponse,
    },
    Command.Stop: {
        TransitType.Request: StopArguments,
        TransitType.Response: StopResponse,
    },
    Command.Status: {
        TransitType.Request: StatusArguments,
        TransitType.Response: StatusResponse,
    },
    Command.StaticStatus: {
        TransitType.Request: StaticStatusArguments,
        TransitType.Response: StaticStatusResponse,
    },
}

File: client/src/bindings.rs
/*!
 * Bindings for the client
 * Created by sheepy0125 | MIT License | 2023-02-19
 */

/***** Setup *****/
// Imports
use crate::events::Event;
use num_derive::FromPrimitive;
use serde::{Deserialize, Serialize};
use std::{fmt::Display, mem::transmute};
use thiserror::Error as ThisError;

// Constants
pub const BAUD_RATE: u32 = 115200_u32;

/***** Error *****/

/// An error from the client
#[derive(ThisError, Debug)]
pub enum ClientError {
    #[error("There was an error with parsing: {0}")]
    Parse(String),
    #[error("There was an error with running: {0}")]
    Run(String),
    #[error("There was an error with the serial connection: {0}")]
    Serial(String),
    #[error("An unknown error occurred: {0}")]
    Unknown(String),
    #[error("{0}")]
    Server(String),
}
impl From<serde_json::Error> for ClientError {
    fn from(value: serde_json::Error) -> Self {
        Self::Parse(value.to_string())
    }
}

/// An error returned by the server
#[repr(u8)]
#[derive(Deserialize, Serialize, Debug, FromPrimitive, Clone, Copy)]
pub enum ServerError {
    MalformedRequestFailedPrefixParsing = 0_u8,
    MalformedRequestFailedCommandParsing = 1_u8,
    MalformedRequestFailedSeparatorParsing = 2_u8,
    MalformedRequestFailedArgumentsParsing = 3_u8,
    MalformedRequestFailedMetadataParsing = 4_u8,
    MalformedRequestTypeError = 5_u8,
    MalformedRequestOtherError = 6_u8,
    Filler7 = 7_u8,
    Filler8 = 8_u8,
    Filler9 = 9_u8,
    MalformedResponseTypeError = 10_u8,
    MalformedResponseOtherError = 11_u8,
    FailedToStartAlreadyStarted = 21_u8,
    FailedToStartMagnetOdometerFailed = 22_u8,
    FailedToStartMotorControlFailed = 23_u8,
    FailedToStopNotStarted = 24_u8,
    FailedToStopStartThreadWouldNotRespond = 25_u8,
    FailedStatusCouldNotAcquireDistanceLock = 26_u8,
    FailedPingNegativeLatency = 27_u8,
    _UpperBound = 28_u8,
    AnyOtherError = 99_u8,
}
impl TryFrom<u8> for ServerError {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        // Ensure not out of bounds
        if value >= Self::_UpperBound as u8 && value != 99_u8 {
            Err(())?;
        }
        // Safety: not out of bounds
        Ok(unsafe { transmute(value) })
    }
}
impl ToString for ServerError {
    fn to_string(&self) -> String {
        match *self as u8 {
            0 => "Malformed request - Failed prefix parsing",
            1 => "Malformed request - Failed command parsing",
            2 => "Malformed request - Failed separator parsing",
            3 => "Malformed request - Failed arguments parsing",
            4 => "Malformed request - Failed metadata parsing",
            5 => "Malformed request - Type error",
            6 => "Malformed request - Other error",
            10 => "Malformed response - Type error",
            11 => "Malformed response - Other error",
            21 => "Failed to start - Already started",
            22 => "Failed to start - Magnet odometer failed",
            23 => "Failed to start - Motor control failed",
            24 => "Failed to start - Could not acquire distance mutex lock",
            25 => "Failed to stop - Not started",
            26 => "Failed to stop - Start thread would not respond",
            27 => "Failed status - Could not acquire distance mutex lock",
            28 => "Failed ping - Negative latency",
            _ => "Any other error",
        }
        .to_string()
    }
}

#[derive(Deserialize, Serialize)]
pub struct ErrorResponse {
    /// This is a u8 for Serde
    pub error_variant: u8,
    pub message: String,
}

/***** Generic bindings *****/

/// Metadata is sent alongside with the request and response
#[derive(Serialize, Deserialize, Debug)]
pub struct MetaData {
    pub time: f64,
}

/// The mode of transit
#[repr(u8)]
#[derive(FromPrimitive)]
pub enum TransitMode {
    ClientToServerRequest = b'?',
    ServerToClientResponse = b'~',
}

/// The type of transit
#[derive(Debug)]
pub enum TransitType {
    Request,
    Response,
}

/***** Commands *****/

#[derive(PartialEq, Eq)]
pub enum Command {
    Ping,
    Start,
    Stop,
    StaticStatus,
    Status,
    Error,
}
impl TryFrom<String> for Command {
    type Error = ClientError; /* Potential type collision */

    fn try_from(value: String) -> Result<Self, ClientError> {
        use Command::*;
        match value.to_ascii_uppercase().as_str() {
            "PING" => Ok(Ping),
            "START" => Ok(Start),
            "STOP" => Ok(Stop),
            "STATICSTATUS" => Ok(StaticStatus),
            "STATUS" => Ok(Status),
            "UNKNOWN" | "ERROR" => Ok(Error),
            _ => Err(ClientError::Parse(format!(
                "Failed to parse command from {value}"
            ))),
        }
    }
}
impl Display for Command {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use Command::*;
        write!(
            f,
            "{}",
            match *self {
                Ping => "PING",
                Start => "START",
                Stop => "STOP",
                StaticStatus => "STATICSTATUS",
                Status => "STATUS",
                Error => "ERROR",
            }
        )
    }
}

/***** Request and response bindings for commands *****/

/// Possible responses
pub enum Response {
    Ping(Event<PingResponse>),
    Start(Event<StartResponse>),
    Stop(Event<StopResponse>),
    Status(Event<StatusResponse>),
    StaticStatus(Event<StaticStatusResponse>),
    Error(Event<ErrorResponse>),
}

// Ping

#[derive(Serialize, Deserialize)]
pub struct PingArguments {
    pub time: f64,
}
#[derive(Deserialize, Serialize)]
pub struct PingResponse {
    pub sent_time: f64,
}

// Start

#[derive(Serialize, Deserialize)]
pub struct StartArguments {
    pub distance: f64,
    pub reverse_brake: bool,
}
#[derive(Deserialize, Serialize)]
pub struct StartResponse;

// Stop
#[derive(Serialize, Deserialize)]
pub struct StopArguments;
#[derive(Deserialize, Serialize)]
pub struct StopResponse;

// Static status

#[derive(Serialize, Deserialize)]
pub struct StaticStatusArguments; //             I don't have the Pi with me so I am
#[derive(Deserialize, Serialize)] //             pretending to be it :)
pub struct StaticStatusResponse {
    pub number_of_magnets: usize,
    pub wheel_diameter: f64,
}

// Regular (dynamic) status

#[derive(Deserialize, Serialize)]
pub struct DistanceInformation {
    /// Centimeters
    pub distance: f64,
    pub velocity: f64,
    pub magnet_hit_counter: usize,
}
#[derive(Serialize, Deserialize)]
pub struct StatusArguments;
#[derive(Deserialize, Serialize)]
pub struct StatusResponse {
    pub running: bool,
    pub uptime: usize,
    pub runtime: usize,
    pub distance: DistanceInformation,
}

/***** Client status *****/

#[repr(u8)]
#[derive(Default, Copy, Clone)]
pub enum ClientStatus {
    #[default]
    GatheringData = 0_u8,
    // Pinging
    SendingPing = 1_u8,
    ReceivingPing = 2_u8,
    // Static status
    RequestingStaticStatus = 3_u8,
    ReceivingStaticStatus = 4_u8,
    // Dynamic status
    RequestingStart = 5_u8,
    ReceivingStatus = 6_u8,
    // Stopping
    RequestingStop = 7_u8,
    Finished = 8_u8,
}
impl ToString for ClientStatus {
    fn to_string(&self) -> String {
        use ClientStatus::*;
        match self {
            GatheringData => "Gathering user input",
            SendingPing => "Pinging (send)",
            ReceivingPing => "Pinging (receive)",
            RequestingStaticStatus => "Getting car information (send)",
            ReceivingStaticStatus => "Getting car information (receive)",
            RequestingStart => "Starting the car (send)",
            ReceivingStatus => "Getting information about the car",
            RequestingStop => "Stopping the car (send)",
            Finished => "Finished",
        }
        .into()
    }
}
impl TryFrom<u8> for ClientStatus {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        if value > Self::Finished as u8 {
            Err(())?;
        }
        Ok(unsafe { transmute((Self::GatheringData as u8) + value) })
    }
}
impl ClientStatus {
    /// If at a boundary, this will return the same thing
    pub fn next(self) -> Self {
        Self::try_from((self as u8) + 1_u8).unwrap_or(Self::Finished)
    }
}
